### PR TITLE
Exercise3

### PR DESCRIPTION
**Pull request by kipmann**
@kipmann  reviewed https://github.com/Systems-Development-and-Frameworks/raum9/pull/4#pullrequestreview-536764210
please review our exercise 3

### CODE REVIEW

File: backend/db.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import crypto from "crypto";
+
+export class User {
+  constructor(data) {
+    Object.assign(this, data);
+  }
+}
+export class Post {
+  constructor(data) {
+    this.id = crypto.randomBytes(16).toString("hex");
```
Comment (JannisBe) [inline backend/db.js:11]: Its rare but possible to have duplicate postids, also very expensive call.
how about removing and adding in addNewPost(postInput)?

File: backend/db.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    super();
+
+    this.usersData = new Set();
+    this.postsData = [];
+  }
+
+  initialize() {}
+
+  addnewPost(postInput) {
+    let post = new Post(postInput);
```
Comment (JannisBe) [inline backend/db.js:30]: How about this?
```suggestion
    let post = new Post({...postInput, id: Math.max(...this.posts.map(post => post.id), 0) + 1});
```

File: backend/server.spec.js [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+  db = new MemoryDataSource();
+  const server = new Server({ dataSources: () => ({ db }) });
+  let testClient = createTestClient(server);
+  query = testClient.query;
+  mutate = testClient.mutate;
+});
+
+describe("queries", () => {
+  describe("USERS", () => {
+    const USERS = gql`
```
Comment (JannisBe) [inline backend/server.spec.js:21]: might be good to declare at the beginning of file for more reusability

File: backend/server.spec.js [hunk 4]
```diff
@@ -0,0 +1,327 @@
+import { jest } from "@jest/globals";
+import { createTestClient } from "apollo-server-testing";
+import { gql } from "apollo-server";
+import Server from "./server";
+import { MemoryDataSource } from "./db";
+
+let query = undefined;
+let mutate = undefined;
+let db = undefined;
```
Comment (JannisBe) [inline backend/server.spec.js:9]: is this needed?

**Review by JannisBe — APPROVED** [review]
nice work :+1:

**Comment by KarakoA** [comment]
@kipmann reviewed https://github.com/Systems-Development-and-Frameworks/raum9/pull/4#pullrequestreview-536764210

**Review by medizinmensch — APPROVED** [review]
Wow, great job!

Great job

You find my comments below. 
I also see your comment that @kipmann reviewed another group. For the review another ⭐ 
Sadly, you missed the opportunity to refactor your repo to the folder structure provided in the exercise. This is the only ❌ you get.

You thus have 9/10 ⭐

**Pull request merged**